File: app/src/test/java/tech/pauly/findapet/discover/AnimalListItemViewModelTest.java
package tech.pauly.findapet.discover;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Arrays;
import java.util.Collections;

import io.reactivex.Single;
import tech.pauly.findapet.data.models.Age;
import tech.pauly.findapet.data.models.Animal;
import tech.pauly.findapet.data.models.Contact;
import tech.pauly.findapet.data.models.Media;
import tech.pauly.findapet.data.models.Photo;
import tech.pauly.findapet.data.models.PhotoSize;
import tech.pauly.findapet.shared.LocationHelper;
import tech.pauly.findapet.shared.ResourceProvider;
import tech.pauly.findapet.shared.datastore.TransientDataStore;
import tech.pauly.findapet.shared.events.ActivityEvent;
import tech.pauly.findapet.shared.events.ViewEventBus;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class AnimalListItemViewModelTest {

    @Mock
    private Animal animal;

    @Mock
    private ViewEventBus eventBus;

    @Mock
    private TransientDataStore dataStore;

    @Mock
    private ResourceProvider resourceProvider;

    @Mock
    private LocationHelper locationHelper;

    @Mock
    private Contact contact;

    private AnimalListItemViewModel subject;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);

        when(animal.getName()).thenReturn("name");
        when(animal.getAge()).thenReturn(Age.ADULT);
        when(animal.getBreedList()).thenReturn(Collections.singletonList("breeds"));
        Media media = mock(Media.class);
        when(media.getPhotoList()).thenReturn(Collections.emptyList());
        when(animal.getMedia()).thenReturn(media);
        when(animal.getContact()).thenReturn(contact);
        when(resourceProvider.getString(Age.ADULT.getFormattedName())).thenReturn("Adult");
        when(locationHelper.getCurrentDistanceToContactInfo(contact)).thenReturn(Single.just(1));
    }

    @Test
    public void onCreate_setAllBasicValuesForInputAnimal() {
        createSubject();

        assertThat(subject.name.get()).isEqualTo("name");
        assertThat(subject.age.get()).isEqualTo("Adult");
        assertThat(subject.breeds.get()).isEqualTo("breeds");
    }

    @Test
    public void onCreate_animalHasMultipleBreeds_setsAllBreeds() {
        when(animal.getBreedList()).thenReturn(Arrays.asList("breed1", "breed2"));

        createSubject();

        assertThat(subject.breeds.get()).isEqualTo("breed1 / breed2");
    }

    @Test
    public void onCreate_getsDistanceForContactInfo() {
        createSubject();

        verify(locationHelper).getCurrentDistanceToContactInfo(contact);
    }

    @Test
    public void onCreate_getsDistanceForContactInfoAndDistanceLessThanZero_distanceVisibilityFalse() {
        when(locationHelper.getCurrentDistanceToContactInfo(contact)).thenReturn(Single.just(-1));

        createSubject();

        assertThat(subject.distanceVisibility.get()).isFalse();
    }

    @Test
    public void onCreate_getsDistanceForContactInfoAndDistanceZero_setsDistanceAndVisibility() {
        when(locationHelper.getCurrentDistanceToContactInfo(contact)).thenReturn(Single.just(0));

        createSubject();

        assertThat(subject.distance.get()).isEqualTo("< 1");
        assertThat(subject.distanceVisibility.get()).isTrue();
    }

    @Test
    public void onCreate_getsDistanceForContactInfoAndDistanceMoreThanZero_setsDistanceAndVisibility() {
        when(locationHelper.getCurrentDistanceToContactInfo(contact)).thenReturn(Single.just(2));

        createSubject();

        assertThat(subject.distance.get()).isEqualTo("2");
        assertThat(subject.distanceVisibility.get()).isTrue();
    }

    @Test
    public void setPhoto_animalHasXPhoto_setsPhoto() {
        Photo photo = mock(Photo.class);
        when(photo.getUrl()).thenReturn("http://url.com");
        when(photo.getSize()).thenReturn(PhotoSize.LARGE);
        Media media = mock(Media.class);
        when(media.getPhotoList()).thenReturn(Collections.singletonList(photo));
        when(animal.getMedia()).thenReturn(media);

        createSubject();

        assertThat(subject.imageUrl.get()).isEqualTo("http://url.com");
    }

    @Test
    public void setPhoto_animalHasNoPhotos_doNotSetPhoto() {
        Media media = mock(Media.class);
        when(media.getPhotoList()).thenReturn(Collections.emptyList());
        when(animal.getMedia()).thenReturn(media);

        createSubject();

        assertThat(subject.imageUrl.get()).isEqualTo("");
    }

    @Test
    public void setPhoto_animalHasNullPhotos_doNotSetPhoto() {
        Media media = mock(Media.class);
        when(media.getPhotoList()).thenReturn(null);
        when(animal.getMedia()).thenReturn(media);

        createSubject();

        assertThat(subject.imageUrl.get()).isEqualTo("");
    }

    @Test
    public void setPhoto_animalDoesNotHaveXPhoto_doNotSetPhoto() {
        Photo photo = mock(Photo.class);
        when(photo.getUrl()).thenReturn("http://url.com");
        when(photo.getSize()).thenReturn(PhotoSize.FEATURED_PET_MODULE);
        Media media = mock(Media.class);
        when(media.getPhotoList()).thenReturn(Collections.singletonList(photo));
        when(animal.getMedia()).thenReturn(media);

        createSubject();

        assertThat(subject.imageUrl.get()).isEqualTo("");
    }

    @Test
    public void setPhoto_mediaIsNull_doNotSetPhoto() {
        when(animal.getMedia()).thenReturn(null);

        createSubject();

        assertThat(subject.imageUrl.get()).isEqualTo("");
    }

    @Test
    public void launchAnimalDetails_launchesAnimalDetails() {
        createSubject();

        subject.launchAnimalDetails();

        verify(eventBus).send(ActivityEvent.build(subject).startActivity(AnimalDetailsActivity.class));
    }

    private void createSubject() {
        subject = new AnimalListItemViewModel(animal, eventBus, dataStore, resourceProvider, locationHelper);
    }
}

File: app/src/test/java/tech/pauly/findapet/discover/AnimalListItemViewModelTest.kt
package tech.pauly.findapet.discover

import io.mockk.*
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import io.reactivex.Observable
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import tech.pauly.findapet.R
import tech.pauly.findapet.data.models.Age
import tech.pauly.findapet.data.models.Animal
import tech.pauly.findapet.data.models.Sex
import tech.pauly.findapet.data.models.Shelter
import tech.pauly.findapet.shared.LocationHelper
import tech.pauly.findapet.shared.ResourceProvider
import tech.pauly.findapet.shared.datastore.AnimalDetailsUseCase
import tech.pauly.findapet.shared.datastore.TransientDataStore
import tech.pauly.findapet.shared.events.ActivityEvent
import tech.pauly.findapet.shared.events.DialogEvent
import tech.pauly.findapet.shared.events.ViewEventBus

class AnimalListItemViewModelTest {

    @MockK
    private lateinit var animal: Animal

    @RelaxedMockK
    private lateinit var eventBus: ViewEventBus

    @RelaxedMockK
    private lateinit var dataStore: TransientDataStore

    @MockK
    private lateinit var resourceProvider: ResourceProvider

    @MockK
    private lateinit var locationHelper: LocationHelper

    @MockK
    private lateinit var contact: Shelter

    @MockK

    private lateinit var subject: AnimalListItemViewModel

    @Before
    fun setup() {
        MockKAnnotations.init(this)
        animal.apply {
            every { id } returns 10
            every { name } returns "name"
            every { age } returns Age.ADULT
            every { formattedBreedList } returns "breeds"
            every { primaryPhotoUrl } returns "photo.jpg"
            every { sex } returns Sex.MISSING
        }

        resourceProvider.apply {
            every { getString(Age.ADULT.formattedName) } returns "Adult"
            every { getString(R.string.pronoun_male_subject) } returns "he"
            every { getString(R.string.pronoun_female_subject) } returns "she"
            every { getString(R.string.pronoun_missing) } returns "it"
            every { getString(R.string.pronoun_male_object) } returns "him"
            every { getString(R.string.pronoun_female_object) } returns "her"
        }

        every { animal.contact } returns contact
        every { locationHelper.getCurrentDistanceToContactInfo(contact) } returns Observable.just(1)
    }

    @Test
    fun onCreate_setAllBasicValuesForInputAnimal() {
        createSubject()

        subject.apply {
            assertThat(id).isEqualTo(10)
            assertThat(name.get()).isEqualTo("name")
            assertThat(age.get()).isEqualTo("Adult")
            assertThat(breeds.get()).isEqualTo("breeds")
            assertThat(imageUrl.get()).isEqualTo("photo.jpg")
        }
    }

    @Test
    fun onCreate_getsDistanceForContactInfoAndDistanceLessThanZero_distanceVisibilityFalse() {
        every { locationHelper.getCurrentDistanceToContactInfo(contact) } returns Observable.just(-1)

        createSubject()

        assertThat(subject.distanceVisibility.get()).isFalse()
    }

    @Test
    fun launchAnimalDetails_launchesAnimalDetailsToDetailsTab() {
        createSubject()

        subject.launchAnimalDetails(AnimalDetailsUseCase.Tab.DETAILS)

        verify { dataStore += AnimalDetailsUseCase(animal, AnimalDetailsUseCase.Tab.DETAILS) }
        verify { eventBus += ActivityEvent(subject, AnimalDetailsActivity::class, false) }
    }

    @Test
    fun showPetWarningDialog_noWarning_launchAnimalDetailsToDetailsTab() {
        createSubject()
        subject = spyk(subject)
        subject.warning.set(false)

        subject.showPetWarningDialog(mockk())

        verify { subject.launchAnimalDetails(AnimalDetailsUseCase.Tab.DETAILS) }
    }

    @Test
    fun showPetWarningDialog_malePet_firesDialogEventWithProperGrammar() {
        every { animal.sex } returns Sex.MALE
        every { resourceProvider.getString(R.string.pet_warning_dialog_body, "name", "him", "he") } returns "name's shelter does not allow for us to search for him directly, so we can\'t tell if he is still available for adoption, sorry!"
        createSubject()
        subject.warning.set(true)
        val slot = slot<DialogEvent>()
        every { eventBus += capture(slot) } answers { nothing }

        subject.showPetWarningDialog(mockk())

        slot.captured.apply {
            assertThat(titleText).isEqualTo(R.string.pet_warning_dialog_title)
            assertThat(bodyText).isEqualTo("name's shelter does not allow for us to search for him directly, so we can\'t tell if he is still available for adoption, sorry!")
            assertThat(positiveButtonText).isEqualTo(R.string.pet_warning_dialog_contact)
            assertThat(negativeButtonText).isEqualTo(R.string.dialog_cancel)
            assertThat(imageUrl).isEqualTo("photo.jpg")
            assertThat(primaryColor).isEqualTo(R.color.warning)
        }
    }

    @Test
    fun showPetWarningDialog_femalePet_firesDialogEventWithProperGrammar() {
        every { animal.sex } returns Sex.FEMALE
        every { resourceProvider.getString(R.string.pet_warning_dialog_body, "name", "her", "she") } returns "name's shelter does not allow for us to search for her directly, so we can\'t tell if she is still available for adoption, sorry!"
        createSubject()
        subject.warning.set(true)
        val slot = slot<DialogEvent>()
        every { eventBus += capture(slot) } answers { nothing }

        subject.showPetWarningDialog(mockk())

        slot.captured.apply {
            assertThat(titleText).isEqualTo(R.string.pet_warning_dialog_title)
            assertThat(bodyText).isEqualTo("name's shelter does not allow for us to search for her directly, so we can\'t tell if she is still available for adoption, sorry!")
            assertThat(positiveButtonText).isEqualTo(R.string.pet_warning_dialog_contact)
            assertThat(negativeButtonText).isEqualTo(R.string.dialog_cancel)
            assertThat(imageUrl).isEqualTo("photo.jpg")
            assertThat(primaryColor).isEqualTo(R.color.warning)
        }
    }

    @Test
    fun showPetWarningDialog_petUnknownGender_firesDialogEventWithProperGrammar() {
        every { animal.sex } returns Sex.UNKNOWN
        every { resourceProvider.getString(R.string.pet_warning_dialog_body, "name", "it", "it") } returns "name's shelter does not allow for us to search for it directly, so we can\'t tell if it is still available for adoption, sorry!"
        createSubject()
        subject.warning.set(true)
        val slot = slot<DialogEvent>()
        every { eventBus += capture(slot) } answers { nothing }

        subject.showPetWarningDialog(mockk())

        slot.captured.apply {
            assertThat(titleText).isEqualTo(R.string.pet_warning_dialog_title)
            assertThat(bodyText).isEqualTo("name's shelter does not allow for us to search for it directly, so we can\'t tell if it is still available for adoption, sorry!")
            assertThat(positiveButtonText).isEqualTo(R.string.pet_warning_dialog_contact)
            assertThat(negativeButtonText).isEqualTo(R.string.dialog_cancel)
            assertThat(imageUrl).isEqualTo("photo.jpg")
            assertThat(primaryColor).isEqualTo(R.color.warning)
        }
    }

    @Test
    fun showPetWarningDialog_clickPositiveButton_launchAnimalDetailsToContactTab() {
        every { animal.sex } returns Sex.UNKNOWN
        every { resourceProvider.getString(R.string.pet_warning_dialog_body, "name", "it", "it") } returns "name's shelter does not allow for us to search for it directly, so we can\'t tell if it is still available for adoption, sorry!"
        createSubject()
        subject.warning.set(true)
        val slot = slot<DialogEvent>()
        every { eventBus += capture(slot) } answers { nothing }
        subject = spyk(subject)

        subject.showPetWarningDialog(mockk())
        slot.captured.positiveButtonCallback?.invoke()

        verify { subject.launchAnimalDetails(AnimalDetailsUseCase.Tab.CONTACT) }
    }

    private fun createSubject() {
        subject = AnimalListItemViewModel(animal, eventBus, dataStore, resourceProvider)
    }
}